### ISSUE TITLE
prefer-const: diagnostic anchored on the `let` declaration instead of the assignment, breaking eslint-disable directives and --report-unused-disable-directives

### ISSUE THREAD

**Issue by ahmedelgabri** [oxc-project/oxc#25630]
## Summary

oxlint's native `prefer-const` anchors its diagnostic on the `let` **declaration** line, whereas ESLint (7 and 9) anchors it on the **assignment/usage** line. This breaks `eslint-disable-next-line prefer-const` directives that were written (correctly, per ESLint semantics) on the line above the assignment: oxlint fires a live error at the declaration and simultaneously reports the assignment-line directive as "unused".

## oxlint version

1.78.0; originally observed on 1.69.0, re-verified on 1.72.0 and 1.78.0.

## Reproduction

`a.ts`:

```ts
export function f(): number {
	let x: number

	// eslint-disable-next-line prefer-const
	x = 1 + 2

	return x
}
```

`.oxlintrc.json`:

```json
{
	"plugins": [],
	"categories": {"correctness": "off"},
	"rules": {
		"prefer-const": ["error", {"destructuring": "any", "ignoreReadBeforeAssign": true}]
	}
}
```

Run:

```
oxlint -c .oxlintrc.json --report-unused-disable-directives a.ts
```

## Expected

ESLint 7 and 9 both report `prefer-const` for `x` (declared with `let`, assigned exactly once, never reassigned), anchored on the **assignment** (`x = 1 + 2`, line 5). The `eslint-disable-next-line prefer-const` directly above it therefore suppresses the report, and `--report-unused-disable-directives` emits nothing (exit 0, no output).

## Actual

oxlint 1.78.0 reports:

```
a.ts:2:6: error eslint(prefer-const): `x` is never reassigned. help: Use `const` instead.
a.ts:4:2: warning: Unused eslint-disable directive (no problems were reported).
```

The diagnostic is anchored on the **declaration** (line 2), not the assignment (line 5), so the directive targeting line 5 does not suppress it — and oxlint then mislabels that directive as unused. A codebase migrating ESLint → oxlint with directives placed per ESLint semantics gets a flood of false "unused directive" warnings plus live errors the directives were meant to suppress.

The same anchor divergence applies to multi-declarator groups assigned later: for `let a: number, b: number` with `a = 1` / `b = 2` on subsequent lines, oxlint anchors both reports on the declarator names (2:6 and 2:17) while ESLint anchors both on the assignment lines (4:2 and 5:2).

(Enabling ESLint's own implementation as a JS plugin — `eslint-js/prefer-const` — anchors the diagnostic on the assignment line, matching ESLint, which confirms the native rule's report location is the divergence, not a config difference.)

## Cross-checked

| Engine | Anchor (fixture without directive) | Directive above assignment |
| --- | --- | --- |
| ESLint 7.23.0 (@typescript-eslint/parser 6.21.0) | 4:2 (assignment) | suppresses; not reported unused |
| ESLint 9.39.4 flat config (typescript-eslint 8.62.1) | 4:2 (assignment) | suppresses; not reported unused |
| oxlint 1.78.0 | 2:6 (`let` declaration) | does not suppress; reported unused |

**Comment by camc314** [comment]
Is this a hard requirement? Why are you disabling the rule in these cases?

I would argue that our diagnostic is clearer because it tells you where the variable is actually defined rather than in some usage of it.

**Comment by ahmedelgabri** [comment]
@camc314 I'm trying to migrate a legacy codebase with lots of these `// eslint-disable` comments. Changing all of them would be a big diff, so I'm trying to minimize the number of changes to get off ESLint, and later I can do the cleanup.
